bitcoin/src/pow.rs

```diff
@@ -913,6 +926,116 @@ impl kani::Arbitrary for U256 {
     }
 }
 
+/// Error from `get_block_header_by_height` callback.
+///
+/// Occurse when there is an error while get a new block header
+/// from an external source.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum GetBlockHeaderError {
+    /// Block Header not found
+    NotFound,
+    /// Generic error with custom message
+    Generic(String),
+}
```

Comment:
Perhaps this would be better
```rust
/// Error from `get_block_header_by_height` callback.
///
/// Occurse when there is an error while get a new block header
/// from an external source.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum GetBlockHeaderError<E> {
    /// Block Header not found
    NotFound,
    /// Custom error.
    Custom(E),
}

impl<E: fmt::Display> fmt::Display for GetBlockHeaderError<E> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        use GetBlockHeaderError::*;

        match *self {
            Custom(ref e) => write_err!(f, "custom error"; e),
            NotFound => f.write_str("block header not found"),
        }
    }
}

#[cfg(feature = "std")]
impl<E: std::error::Error + 'static> std::error::Error for GetBlockHeaderError<E> {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        use GetBlockHeaderError::*;

        match *self {
            Custom(ref e) => Some(e),
            NotFound => None,
        }
    }
}
```
Can you put all the error stuff _below_ where it is used i.e., below `calculate_next_work_required`

Reply:
yes, I was thinking about this yesterday, thanks for writing the code. Apply this